gdexlab/tcp-load-balancer
# Least Connections Request Forwarder

## What?
This PR adds all of the setup needed for the load balancer to run locally, and forward traffic upstream based on least connections. 

The intent is to satisfy 2/6 requirements from the [gravitational career challenge](https://github.com/gravitational/careers/blob/rjones/challenge-2.md/challenges/systems/challenge-2.md). 

- **Implement a least connections request forwarder that tracks the number of connections per upstream.**
- **Accept and forward requests to upstreams.**

The files included cover the [Request Forwarder](https://github.com/gdexlab/tcp-load-balancer/blob/main/rfd/001-tcp-load-balancer-design-doc.md#request-forwarder) (excepting the authn/authz portions)  and [Demonstrating Functionality](https://github.com/gdexlab/tcp-load-balancer/blob/main/rfd/001-tcp-load-balancer-design-doc.md#demonstrating-functionality) sections of the design doc.

This will likely be the largest of the PRs on this project, since it contains much of the application setup and structure.

## How to Test It?
Run unit tests with `go test ./...` from the root of the repository.

Use `go run main.go -p 50043` to start the load balancer locally and have it listen on port `50043`.  By default, the current number of static hosts will be 5 with 7 static clients. Those can be modified in the `internal/config/config.go` file. 

You will notice log statements from the static clients passing data concurrently through the load balancer to upstream hosts. An example of the log output is attached:
```
▶ go run main.go -p 50043
2022/06/21 22:51:28 Load balancer listening on [::]:50043
2022/06/21 22:51:28 Statically defined host listening on [::]:57399
2022/06/21 22:51:28 Statically defined host listening on [::]:57400
2022/06/21 22:51:28 Statically defined host listening on [::]:57401
2022/06/21 22:51:28 Statically defined host listening on [::]:57402
2022/06/21 22:51:28 Statically defined host listening on [::]:57403
2022/06/21 22:51:32 Static client at [::1]:57407 sending 'Hello' to LB address: [::1]:50043
2022/06/21 22:51:32 Static client at [::1]:57405 sending 'Hello' to LB address: [::1]:50043
2022/06/21 22:51:32 Static client at [::1]:57408 sending 'Hello' to LB address: [::1]:50043
2022/06/21 22:51:32 Static client at [::1]:57410 sending 'Hello' to LB address: [::1]:50043
2022/06/21 22:51:32 Static client at [::1]:57409 sending 'Hello' to LB address: [::1]:50043
2022/06/21 22:51:32 Static client at [::1]:57406 sending 'Hello' to LB address: [::1]:50043
2022/06/21 22:51:32 Static client at [::1]:57411 sending 'Hello' to LB address: [::1]:50043
2022/06/21 22:51:32 Client at [::1]:57409 received response: Data 'Hello' was received by host at [::]:57399 
2022/06/21 22:51:32 Client at [::1]:57410 received response: Data 'Hello' was received by host at [::]:57402 
2022/06/21 22:51:32 Client at [::1]:57405 received response: Data 'Hello' was received by host at [::]:57399 
2022/06/21 22:51:32 Client at [::1]:57407 received response: Data 'Hello' was received by host at [::]:57400 
2022/06/21 22:51:32 Client at [::1]:57411 received response: Data 'Hello' was received by host at [::]:57400 
2022/06/21 22:51:32 Client at [::1]:57406 received response: Data 'Hello' was received by host at [::]:57401 
2022/06/21 22:51:32 Client at [::1]:57408 received response: Data 'Hello' was received by host at [::]:57403
...
```

Additionally you may open a tcp connection (after the app has started up) using a command like `telnet localhost 50043` to send data through the load balancer. Exit the telnet prompt with `Ctrl + ]` and then type `quit`.
